affine centered on the ROI

jlaura opened the pull request:
This PR alters the subpixel registration to use an affine transform that is centered on the ROI (instead of the upper left corner). I have a notebook that demos how this is working that I would be happy to show via a private gist link. 

I will comment inline if that helps though I tried to comment in the code directly so that the why is longer lived.

jlaura commented:
Some debugging print statements remain. I will work to pull those. Wanted to get this up in case anyone wanted to test locally.

Kelvinrr commented on autocnet/matcher/subpixel.py:
```diff
@@ -600,57 +689,65 @@ def geom_match(base_cube,
         metric = (maxcorr, perror, pdiff)
     else:
-        x,y,maxcorr,temp_corrmap = restemplate
+        x,y,metric,temp_corrmap = restemplate
```
I feel like we might as well remove the phase_kwargs options, still seems half implemented here (e.g. no error checking to make sure subpixel template returns even enable a good phase match).

jlaura replied:
I would be down with fully yanking phase. It hasn't been consistently effective.

ladoramkershner replied:
I will third that motion

Kelvinrr commented on autocnet/matcher/subpixel.py:
```diff
@@ -600,57 +689,65 @@ def geom_match(base_cube,
-            axs[1].imshow(bytescale(dst_arr[2:]), cmap="Greys_r")
-            axs[1].scatter(size_x, size_y, s=10, color='red')
-            axs[1].plot([0, size_x*2], [size_y, size_y], linewidth=0.75, color='red')
-            axs[1].plot([size_x, size_x], [0, size_y*2], linewidth=0.75, color='red')
-            axs[1].set_xlim([0, dst_arr[2:].shape[0]])
-            axs[1].set_ylim([0, dst_arr[2:].shape[1]])
+            #axs[0].set_xlim([0, base_arr.shape[0]])
+            #axs[0].set_ylim([0, base_arr.shape[1]])
+            axs[1].set_title("Unprojected Image\n w/ original point")
+            dst_arr = input_cube.read_array(pixels=dst_pixels, dtype=dst_type)
+            axs[1].imshow(bytescale(dst_arr), cmap="Greys_r")
+            #axs[1].scatter(warped_center_x, warped_center_y, s=10, color='blue')
+            nx, ny = affine([x,y])[0]
+            ox, oy = affine(warped_center)[0]
+            print(nx, ny)
+            axs[1].scatter(nx, ny, s=10, color='red')
+            axs[1].scatter(ox, oy, s=10, color='blue')
+            #axs[1].plot([0, size_x*2], [size_y, size_y], linewidth=0.75, color='red')
+            #axs[1].plot([size_x, size_x], [0, size_y*2], linewidth=0.75, color='red')
+            #axs[1].set_xlim([0, dst_arr[2:].shape[0]])
+            #axs[1].set_ylim([0, dst_arr[2:].shape[1]])
             axs[2].set_title("Projected Image\n w/ registered point")
             axs[2].imshow(bytescale(dst_arr[2:]), cmap="Greys_r")
             axs[2].scatter(x, y, s=10, color='red')
             axs[2].plot([0, size_x*2], [y, y], linewidth=0.75, color='red')
             axs[2].plot([x, x], [0, size_y*2], linewidth=0.75, color='red')
-            axs[2].set_xlim([0, dst_arr[2:].shape[0]])
-            axs[2].set_ylim([0, dst_arr[2:].shape[1]])
+            #axs[2].set_xlim([0, dst_arr[2:].shape[0]])
+            #axs[2].set_ylim([0, dst_arr[2:].shape[1]])
 
-            plt.show()
+            plt.show()"""
```
I assume this is being moved to subpixel template?

jlaura replied:
Exactly. I need to add a verbose option pass through into that function. I think that is probably a better place for that vis to happen as it is closer to the actual place where this stuff is used.

Kelvinrr reviewed (commented):
Can you message me the gist link? probably try it out sometime tomorrow to see how well it works. Is this a draft because you're not confident in the logic or is it simply clean-up still required?

jlaura commented:
And still a draft because I would appreciate more eyes on the logic and I want to clean up the prints/debugs/etc.

ladoramkershner commented on autocnet/matcher/cross_instrument_matcher.py:
```diff
@@ -105,8 +105,12 @@ def generate_ground_points(Session, ground_mosaic, nspts_func=lambda x: int(roun
-
+ 
+        
         linessamples = isis.point_info(ground_mosaic.file_name, p.x, p.y, 'ground')
+        if linessamples is None:
+            print('unable to find point in ground image')
+            continue
```
I think this should be wrapped in a try/except. point_info is set up to throw exceptions if there is an issue, I am not sure if it can return None. Did you run into None returns?

jlaura replied:
Yup. It was throwing an error and returning None. So I went with the message.

ladoramkershner commented on autocnet/matcher/subpixel.py:
```diff
@@ -227,6 +232,7 @@ def subpixel_template(sx, sy, dx, dy,
                       image_size=(251, 251),
                       template_size=(51,51),
                       func=pattern_match,
+                      transform = None,
```
Add a description to function doc string

jlaura commented:
I just updated:

1. Removed the phase stuff
1. Cleaned up prints
1. Cleaned up the verbose stuff, located primarily in subpixel. Default is False so that we maintain current functionality.
1. Added doc string entries where they were missing.

If this looks good to people and they have tested to their satisfaction locally, I am comfortable converting the PR from draft status.

ladoramkershner commented:
I pulled it locally and it looks good, there is just a lot of crap ground getting through because we have no threshold on the original propagation up from THEMIS to CTX. Actually no threshold kwargs in propagate_cnet at all, I believe I was originally against adding that kwargs, but idk why. 

Also, I think we could benefit from having an error similar to ISIS where, if the max correlation value is on the edge of the fit chip, an 'Insufficient Data' error is thrown and that match is thrown out.

jlaura commented:
@ladoramkershner I can add both of those things.

ladoramkershner commented:
I think that capability is off topic for this PR, which already rather large. I am going to open a ticket and we can address it in a separate PR down the line.

jlaura commented:
@ladoramkershner Last CI test is running now and I added tests to increase total test coverage. This should be good for another review whenever.

ladoramkershner commented on autocnet/matcher/cross_instrument_matcher.py:
```diff
@@ -259,10 +263,12 @@ def propagate_point(Session,
         sx, sy = m["sample"], m["line"]
 
         for i,image in images.iterrows():
+            print('PATH', image['path'])
+            # When grounding to THEMIS the df has a PATH to the QUAD
```
possible print clean up? Not necessary if you really think it is valuable here.

jlaura replied:
Removed.

ladoramkershner commented on autocnet/matcher/subpixel.py:
```diff
@@ -271,26 +430,56 @@ def subpixel_template(sx, sy, dx, dy,
+    try:
+        d_template_dtype = isis2np_types[pvl.load(d_img.file_name)["IsisCube"]["Core"]["Pixels"]["Type"]]
+    except:
+        d_template_dtype = None
```
What happens if this goes through the exception and we try and load s_image and d_template with None dtypes?

jlaura replied:
The default is None once it winds all the way through to plio where it does the read. The reality is that this should end up in plio at some point and be removed from here.

ladoramkershner commented on autocnet/matcher/subpixel.py:
```diff
@@ -271,26 +430,56 @@ def subpixel_template(sx, sy, dx, dy,
+    transformed_roi = d_template
+
+    if verbose:
+        axs[2].imshow(transformed_roi, cmap='Greys')
+        axs[3].imshow(d_template[5:-5,5:-5], cmap='Greys')
```
These are both equal to d_template and that has already been plotted. Are these plots useful?

ladoramkershner replied:
I think these plots may be useful in subpixel_transformed_template, but replace the 5's for template_buffer kwarg

jlaura commented:
@ladoramkershner Updated based on your comments.

ladoramkershner commented on autocnet/matcher/subpixel.py:
```diff
@@ -546,129 +760,25 @@ def geom_match(base_cube,
-    return sample, line, dist, metric, temp_corrmap
+
+    # Estimate the transformation
+    affine = estimate_affine_transformation(base_corners, dst_corners)
```
The function call for this is: estimate_affine_transformation(destination_coordinates, source_coordinates)

That seems like the opposite of how you are calling it. I remember looking at you result and they looked good (especially since when an affine transformation is used in the non intended direction it is pretty obvious), but it terms of documentation and technical correctness, I think either this line (and the corresponding down stream code) should be swapped, or the estimate_affine() could be swapped?

jlaura replied:
The naming here sucks. So I am going to address the really good comment by making the naming consistent with skimage.

ladoramkershner commented:
I finished reviewing. I think my two comments are:
1.) Resolving the function signature on estimate_affine_transformation (or explaining to me why it is fine as is)
2.) Moving the plots that were in subpixel_template into subpixel_transformed_template and removing/adding the 'verbose' kwargs accordingly. I really liked those plots!

Always encourage push back if you don't agree, I just wanted to sum up my review!

jlaura commented:
@ladoramkershner This is updated based on the always appreciated comments. I updated the figures for the transformed work as well to include titles.

![Screen Shot 2020-09-25 at 1 50 06 PM](https://user-images.githubusercontent.com/2105468/94314795-1ee71900-ff36-11ea-820f-77fc8a475b43.png)

jlaura commented:
I have an idea how to also avoid high correlations near the edge of the template. Once this PR looks good and is merged I will open a PR with that modification.

ladoramkershner commented on autocnet/matcher/subpixel.py:
```diff
@@ -645,10 +655,10 @@ def geom_match(base_cube,
     ----------
-    base_cube:  plio.io.io_gdal.GeoDataset
+    destination_cube:  plio.io.io_gdal.GeoDataset
                 source image
 
-    input_cube: plio.io.io_gdal.GeoDataset
+    sourcecube: plio.io.io_gdal.GeoDataset
                 destination image; gets matched to the source image
```
These seem backwards..

jlaura replied:
Updated the doc string.

ladoramkershner commented on autocnet/matcher/subpixel.py:
```diff
@@ -703,76 +713,76 @@ def geom_match(base_cube,
     # Can we validate this inside the ROI object?
-    base_size_x = template_kwargs['image_size'][0]
-    base_size_y = template_kwargs['image_size'][1]
+    destination_size_x = template_kwargs['image_size'][0]
+    destination_size_y = template_kwargs['image_size'][1]
```
This also seems backwards. The image is the larger out of template and image chips. So you are trying to find out how template moves to line up with image. So (I am thinking in terms of ground matching) image -> Themis -> source

jlaura replied:
This is correct. The image_size kwarg is not named the best, but needs to be consistent with sub pixel_template naming conventions. The destination is Themis. We are going to warp the source, CTX, into the destination.
